File: Leetcode/LargestDivisibleSubset.rs
// https://leetcode.com/problems/largest-divisible-subset

struct Solution {}

impl Solution {
    pub fn largest_divisible_subset(mut nums: Vec<i32>) -> Vec<i32> {
        let n = nums.len();
        let mut dp = vec![(-1, 0, 1); n];

        nums.sort();
        for i in 0..n {
            dp[i].1 = i;
        }
        for (i, x) in nums.iter().enumerate() {
            for (j, y) in nums.iter().enumerate().skip(i + 1) {
                if *y % *x == 0 && dp[j].2 < dp[i].2 + 1 {
                    dp[j].2 = dp[i].2 + 1;
                    dp[j].0 = i as i32;
                }
            }
        }

        let mut cur = dp.iter().max_by(|a, b| a.2.cmp(&b.2)).unwrap();
        let mut ans = vec![];

        loop {
            ans.push(nums[cur.1 as usize]);

            if cur.0 == -1 {
                ans.reverse();
                break ans;
            }

            cur = &dp[cur.0 as usize];
        }
    }
}

File: Leetcode/LargestDivisibleSubset.c
// https://leetcode.com/problems/largest-divisible-subset

#define N 1000

int comp(const void* a, const void* b) {
    int* ai = (int*)a;
    int* bi = (int*)b;

    return *ai - *bi;
}

int* largestDivisibleSubset(int* nums, int numsSize, int* returnSize) {
    qsort(nums, numsSize, sizeof(int), comp);

    int dp[N];
    int par[N];

    for (int i = 0; i < numsSize; ++i) {
        dp[i] = 1;
        par[i] = i;
    }

    int ma_i = 0;
    for (int i = 0; i < numsSize; ++i) {
        for (int j = i + 1; j < numsSize; ++j) {
            if (nums[j] % nums[i] == 0 && dp[j] < dp[i] + 1) {
                dp[j] = dp[i] + 1;
                par[j] = i;
            }

            if (dp[ma_i] < dp[j]) {
                ma_i = j;
            }
        }
    }

    *returnSize = dp[ma_i];
    int* ans = (int*)malloc(sizeof(int) * dp[ma_i]);
    while (true) {
        ans[dp[ma_i] - 1] = nums[ma_i];

        if (ma_i == par[ma_i]) {
            break;
        }
        ma_i = par[ma_i];
    }

    return ans;
}